### PR TITLE
uparrow copies last message to compose

### LINKED ISSUE #20: Pressing up in compose should display last message
n/t

### PR DESCRIPTION
**Pull request by varsnotwars**
Behaviour:

- ArrowUp: begins cycling through the 10 most recent messages if there are any. Pressing ArrowUp/ArrowDown once this has started cycles through recent messages, it will wrap around once you hit the last element in each direction.

- ArrowUp/ArrowDown (when the focus is on the identities select dropdown, after changing an identity, for example): toggles the identities, pressing any other key from here will bring the focus to the text box. I believe this is expected default behaviour? I can change this if not.

- ArrowDown (when the focus is anywhre other than the identities select dropdown): brings the focus to the text box.

Note:
Currently messages that have identical text are filtered out. This was better UX when the identical messages were one after the other as it appears the message is not updating. 

closes #20

### CODE REVIEW

**Comment by indutny** [comment]
Hello!

Thank you very much for making this Pull Request.

This is very close to what we are looking for. I'm going to leave a few comments with explanations.

File: src/components/Compose.js [hunk 1]
```diff
@@ -46,12 +47,16 @@ function Compose(props) {
   const input = useRef();
 
   useEffect(() => {
-    const onKeyDown = (e) => {
+    const onKeyDown = e => {
```
Comment (indutny) [inline src/components/Compose.js]: Nitpicking: unnecessary change.
```suggestion
    const onKeyDown = (e) => {
```

File: src/components/Compose.js [hunk 2]
```diff
@@ -22,6 +22,7 @@ function Compose(props) {
     onBeforePost,
     postFile,
     addNotification,
+    lastMessageText,
```
Comment (indutny) [inline src/components/Compose.js]: I think it should live in `state`, and `updateState` should be used for storing the last message.

File: src/components/Compose.js [hunk 3]
```diff
@@ -46,12 +47,16 @@ function Compose(props) {
...
   useEffect(() => {
-    const onKeyDown = (e) => {
+    const onKeyDown = e => {
       if (!e.key || e.metaKey || e.ctrlKey) {
         return;
       }
 
       if (input.current && !isPickerVisible) {
+        // don't overwrite if something already typed
+        if (e.keyCode === 38 && !message) {
```
Comment (indutny) [inline src/components/Compose.js]: This has been a big surprise for me too, but [keyCode](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) is deprecated. Could you use `.code` here? (Not sure if it implies changing the `38` to something else.

File: src/pages/Channel.js [hunk 4]
```diff
@@ -129,6 +129,12 @@ function Channel(props) {
     </span>;
   }
 
+  let lastMessage = channel.messages[channel.messages.length - 1];
```
Comment (indutny) [inline src/pages/Channel.js]: This is nice, but I think UX-wise people would expect to get their last message here. It doesn't have to come from the channel history, though.

The way I imagine it (and I think the way it works in most IRC clients) is that there is a short array of the messages (say last 10 messages) that **you** sent on the channel, and you can change between them with up and down arrows.

**Comment by varsnotwars** [comment]
Hey, @indutny, sorry for all the email spam 😬.

So I've implemented what you asked in your comments, including cycling through the last 10 messages with arrow keys. Up arrow starts, and then up/down keys cycles through. It wraps around when you pass the start/end.

Also there is one last issue I'm having:

So the way I get the recent messages for the current identity is using the pubKey that is attached to the message's author. This works for my identities that are in a channel that identity did not create. e.g.,
` identity A in channel A does not have their pubKey attached to their messages. But Identity B in channel A, does have their pubKey attached to their messages.`

 Is this a bug? I created both identities but the pubKey is still missing in the above scenarios.

Could you review the code and test it locally, please.

To test the work just use an identity that was invited to the channel, not the owner identity.

**Comment by varsnotwars** [comment]
Updated description with extra info.

File: src/components/Compose.js [hunk 5]
```diff
@@ -51,8 +129,30 @@ function Compose(props) {
...
+              setMessage(nextMessage);
+            }
+          }
+          input.current.focus();
+        } else if(isSelect && !isUpOrDown) { // so non up/down keys bring focus to textarea
+          input.current.focus();
+        }
       }
     };
     document.addEventListener('keydown', onKeyDown);
```
Comment (indutny) [inline src/components/Compose.js]: I'm slightly concerned about this having a global effect. I think it would be better if the `keydown` for history lookup would be listened only on the `<input/>` itself, and not for the whole document

File: src/components/Compose.js [hunk 6]
```diff
@@ -83,6 +183,10 @@ function Compose(props) {
     setIdentityKey(identities[0].publicKey);
   }
 
+  if (messages.length && identityKey) {
+    setUsersRecentMessages(identityKey);
```
Comment (indutny) [inline src/components/Compose.js]: What do you think about taking a way way simpler route here: instead of looking up the history from displayed messages (which has its pros!), we could just track the history since the app start. In other words, every posted message (see `postMessage` in `src/redux/actions.js`) would dispatch a redux event to add a message to the history of the channel. This way we save some execution time here without sacrificing much of usability.

Comment (indutny) [reply src/components/Compose.js]: I'd probably add a `history` field somewhere [here](https://github.com/peerlinks/peerlinks-desktop/blob/master/src/redux/reducers.js#L337) and add another action at the big switch in `channels` in the same file. The action type could be something like `CHANNEL_PUSH_TO_HISTORY`, and the payload would obviously be the message text itself.

**Comment by varsnotwars** [comment]
@indutny, I agree with your comments and made those changes. One extra change I've made is tracking the index of the current recent message. This is needed to handle identical message text.
